java/rocksjni/associative_merge_operator_jnicallback.cc

```diff
@@ -0,0 +1,127 @@
+    if (returnTypeInitMethodId == 0)
+      throwJavaLangError(env, "unable to find field ReturnType.<init>");
+
+    returnTypeFieldId = env->GetFieldID(returnTypeClass, "isArgumentReference", "Z");
+    if (returnTypeFieldId == 0)
+      throwJavaLangError(env, "unable to find field ReturnType.isArgumentReference");
+    catchAndLog(env);
+  }
+
+  bool AssociativeMergeOperatorJniCallback::Merge(const Slice &key,
+                                                  const Slice *existing_value,
+                                                  const Slice &value,
+                                                  std::string *new_value,
+                                                  Logger* /*logger*/) const {
+    jboolean attached_thread = JNI_FALSE;
+    JNIEnv *env = getJniEnv(&attached_thread);
+    if (env == NULL)
+      return false;
+
+    jbyteArray jb0, jb1, jb2;
+    jbyte *buf0;
+    jbyte *buf1;
+    jbyte *buf2;
+
+    size_t _s0 = key.size() * sizeof(char);
+    const jsize s0 = static_cast<jsize>(_s0);
+    buf0 = (jbyte *) key.data();
+    jb0 = env->NewByteArray(s0);
+    env->SetByteArrayRegion(jb0, 0, s0, buf0);
+
+    if (existing_value != NULL) {
+      size_t _s1 = existing_value->size() * sizeof(char);
+      const jsize s1 = static_cast<jsize>(_s1);
+      buf1 = (jbyte *) existing_value->data();
+      jb1 = env->NewByteArray(s1);
+      env->SetByteArrayRegion(jb1, 0, s1, buf1);
+    } else {
+      buf1 = NULL;
+      jb1 = NULL;
+    }
+
+    size_t _s2 = value.size() * sizeof(char);
+    const jsize s2 = static_cast<jsize>(_s2);
+    buf2 = (jbyte *) value.data();
+    jb2 = env->NewByteArray(s2);
+    env->SetByteArrayRegion(jb2, 0, s2, buf2);
+
+    jobject rtobject = env->NewObject(returnTypeClass, returnTypeInitMethodId);
+    jbyteArray jresult = (jbyteArray) env->CallObjectMethod(m_jcallback_obj,
+                                                            mergeMethodId,
+                                                            jb0, jb1, jb2, rtobject);
+    jthrowable ex = env->ExceptionOccurred();
+    catchAndLog(env);
+    env->DeleteLocalRef(jb0);
+    catchAndLog(env);
+    if (jb1 != NULL)
+      env->DeleteLocalRef(jb1);
+    env->DeleteLocalRef(jb2);
+    catchAndLog(env);
+    jboolean rtr = env->GetBooleanField(rtobject, returnTypeFieldId);
+    env->DeleteLocalRef(rtobject);
+
+    if (ex) {
+      if (jresult != nullptr) {
+        env->DeleteLocalRef(jresult);
+      }
+      env->Throw(ex);
+      releaseJniEnv(attached_thread);
+      return false;
+    } else {
+      int len = env->GetArrayLength(jresult) / sizeof(char);
+      char *result = (char *) env->GetByteArrayElements(jresult, 0);
+      new_value->clear();
+      new_value->assign(result, len);
+      env->ReleaseByteArrayElements(jresult, (jbyte *) result, rtr ? JNI_COMMIT : JNI_ABORT);
+      env->DeleteLocalRef(jresult);
+      releaseJniEnv(attached_thread);
+      return true;
+    }
+  }
```

Comment:
In general we need a bit more error checking above around the JNI calls.

Reply:
I add a lot of checks around jni calls